[WIP] Add component-level dirty/changed flags

JoNax97 opened the pull request:
Adds the ability to flag entities as dirty on a per-component basis. This can be done manually by calling `entity.SetDirty<MyComponentType>()` or (coming soon) automatically when a query potentially modifies it.


It works by storing a BitSet for each component type inside each chunk. 

It's behind a compile-time flag and even when enabled, it should have minimal cpu and memory overhead when unused.

Future work:

- [ ] Finish dirty-filtering in queries
- [ ] Add automatic dirtying to queries that write to a component
- [ ] Add a single chunk-level flag to speed up dirty checking

genaray commented:
@JoNax97 
Thanks! 
Looks fine to me, however i wonder whats the Performance Impact will be If every entity hast its own bitset stored in the chunk. Wondering If there better ways of doing that 🤔

genaray commented:
> I considered making the change tracker a component, but in order to avoid unexpected structural changes, the user would need to opt-in to the feature at entity creation time. What's your view on that?
> 
> Additionally, it would make it less performant to check for changed components in queries. The current approach wil allow for skipping whole chunks with a single boolean check

True, I assume this is the best way of doing it so far. I guess we could merge it then, tests are missing tho :)

JoNax97 commented:
Updated the PR with most of the support for filtering non-changed entities in queries. For simple cases where we need to iterate over the entity or a single component only, a new enumerator encapsulates all the changed logic while keeping the same performance as before.

For more complex cases, a single iterator can't return multiple values in a performant way. I will address those by injecting the new logic directly in the loops. 

Some methods (like query-based add, set, remove) will not support change filtering, so I've added an assertion to reject such queries. This is to avoid confusion, as otherwise the query will affect more entities than expected.

I also added a bunch of new tests, with different combinations of filters.

I still have some work to do but the core it's already in place. I'd appreciate any feedback you have
